bip-0345.mediawiki

```diff
@@ -0,0 +1,683 @@
+** If this value does not decode to a valid CScriptNum, script execution when spending this output MUST fail and terminate immediately.
+** If this value is greater than or equal to the number of outputs, script execution when spending this output MUST fail and terminate immediately.
+** If this value is negative and not equal to -1, script execution when spending this output MUST fail and terminate immediately.<ref>'''Why is -1 the only allowable negative value for revault-vout-idx?''' A negative revault index indicates that no revault output exists; if this value were allowed to be any negative number, the witness could be malleated (and bloated) while a transaction is waiting for confirmation.</ref>
+
+* <code><revault-amount></code> is an up to 7-byte CScriptNum-encoded number indicating the number of satoshis being revaulted.
+** If this value does not decode to a valid CScriptNum, script execution when spending this output MUST fail and terminate immediately.
+** If this value is not greater than or equal to 0, script execution when spending this output MUST fail and terminate immediately.
+** If this value is non-zero but <code><revault-vout-idx></code> is negative, script execution when spending this output MUST fail and terminate immediately.
+
+After the stack is parsed, the following validation checks are performed:
+
+* Let the output designated by <code><trigger-vout-idx></code> be called ''triggerOut''.
+* If the scriptPubKey of ''triggerOut'' is not a witness program of the same version and same tapleaf version as the currently executing script, script execution MUST fail and terminate immediately.
+* Let the script constructed by taking the <code><leaf-update-script-body></code> and prefixing it with minimally-encoded data pushes of the <code><n-pushes></code> leaf-update script data items be called the ''leaf-update-script''.
+** Note: the leaf-update data items will be in the same order in the ''leaf-update-script'' as they appeared on the stack.
+* If the scriptPubKey of ''triggerOut'' does not match that of a taptree that is identical to that of the currently evaluated input, but with the leaf script substituted for ''leaf-update-script'', script execution when spending this output MUST fail and terminate immediately.
+** Note: the parity bit of the resulting taproot output is allowed to vary, so both values for the new output must be checked.
+* Let the output designated by <code><revault-vout-idx></code> (if the index value is non-negative) be called ''revaultOut''.
+* If the scriptPubKey of ''revaultOut'' is not equal to the scriptPubKey of the input being spent, script execution when spending this output MUST fail and terminate immediately.
+* Implemetation recommendation: if the sum of the amounts of ''triggerOut'' and ''revaultOut'' (if any) are not greater than or equal to the value of this input, script execution when spending this output SHOULD fail and terminate immediately.
```

Comment:
I think this could use additional clarification. On initial read, I was confused as to why the ''triggerOut'' amount should be greater than or equal to the input amount. To this point, the BIP discusses how once vaulted, an input can be triggered for withdrawal. Fees were glanced over in the beginning. So my mental model was that an input in a vault would be spent through a trigger tx as follows:
> 1. Specify input to be spent (input_idx) which contains an amount (A)
> 2. Specify trigger amt (B)
> 3. Specify revault amt (C)
> 4. Spend input_idx with amt A such that: fee = A - (B + C)

In this case, the logic would be that the sum of ''triggerOut'' and ''revaultOut'' must be less than or equal to the value of the input for this tx to be valid. While some clarification is made later in the BIP, I think a discussion of the fee structure needs to happen before walking through the Python code of trigger checks.

Reply:
fees aren't a part of the specification, "protecting input value" is. It's an accumulator of value-to-outputs over the various OP_VAULT invocations. 

that said, if you can think of a way to make the text clearer, would be great.

Reply:
"implementation" has two n's in it